docs/okeps/okep-5259-no-overlay.md

```diff
@@ -0,0 +1,598 @@
+
+The following diagram shows the OVN topology of a cluster with three networks:
+
+* a CUDN (in green) that uses Geneve encapsulation (default behavior)
+* a default network (in gray) that enables no-overlay mode
+* a CUDN (in blue) that enables no-overlay mode
+
+![No-overlay topology](../images/no_overlay_topology.jpg)
+
+#### No SNAT in no-overlay mode for egress traffic
+
+The BGP feature in ovn-kubernetes only disables SNAT for traffic that is
+destined to external networks, not for traffic that is destined to other pods in
+the cluster.
+
+In no-overlay mode, we want to fully leverage BGP and avoid the SNAT step on the
+gateway router for intra-cluster traffic as well. Pod-to-pod traffic within the
+same network will hit the BGP-learned routes to reach its destination node,
+without being encapsulated in a Geneve tunnel and without being SNATed to the
+node IP.
+
+The only exception is for pod-to-other_node traffic, to ensure nodePort services
+can be accessed across networks. The egress traffic will be SNATed to the node
+IP before leaving the node.
+
+#### Isolation between UDNs with no-overlay mode enabled
+
+The [existing UDN isolation
+mechanism](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5186/commits/8f6e7d30ee5f4926a21e2de75488aad80344814b)
+will still be in place. Pods in different UDNs will be isolated from each other
+and from the default network by means of the existing OVN ACLs on every node
+logical switch: ACLs are enforced on outgoing traffic at each advertised UDN
+switch, verifying whether both source and destination IPs belong to the same
+advertised UDN. If the destination IP does not come from the same UDN subnet as
+the source IP, traffic is dropped. The no-overlay mode will not affect the
+isolation between UDNs or between UDNs and the default network.
+
+### Workflow
+
+#### Enable No-overlay Mode for the Default Network
+
+No-overlay mode for the default network must be enabled before OVN-Kubernetes
+pods start. Therefore, the following configuration manifests shall be created
+before deploying the OVN-Kubernetes pods.
+
+1. The cluster admin enables no-overlay mode by creating the well-known
+   `default` UDN CR with `spec.layer3.encapsulation` set to `None`.
+
+    ```yaml
+    apiVersion: k8s.ovn.org/v1
+    kind: ClusterUserDefinedNetwork
+    metadata:
+      # 'default' is the reserved name to represent the cluster default network
+      name: default
+    spec:
+      # use an empty selector to select all namespaces
+      namespaceSelector: {}
+      network:
+        topology: Layer3
+        layer3:
+          role: Primary
+          mtu: 1500
+          subnets:
+          - cidr: 10.244.0.0/16
+            hostSubnet: 24
+          encapsulation: None
+    ```
+
+1. OVN-kubernetes cluster manager watches the `default` CUDN CR. If
+   `spec.layer3.encapsulation` equals `None`, cluster manager will create a
+   `RouteAdvertisement` CR which will advertise pod subnets in the default
```

Comment:
So this RouteAdvertisements would be owned by ovnk. If this is the case then I think it is worthwhile to have some implementation details on how this is reconciled and how this status is reflected back to the CUDN crd status. I ask myself questions like: what happens if someone deletes this RouteAdvertisements?

Reply:
On the other side, why does it it need to be created/owned by OVNK? Why isn't the user responsible for it?
Or, why don't we orchestrate no overlay mode for the default network from CNO? It might make sense to create the necessary FRRConfigurations, RouteAdvertisments and pass appropriate arguments to OVNK from there.

Reply:
yeah I'm very wary of automatically creating this CR. I'd prefer for default network we just use a config flag. Unless you can tell us more about the plans for dynamically turning it on and off for default network in the future. I would argue it probably makes sense to force the user to re-roll out the ovnk daemonset, since this is like a one time config change that isnt going to change often.

Reply:
To handle MTU issues, we have the MTU migration process. That would be enough for the default network. At some point we probably would need to find a way to loop in the UDNs in that MTU migration process, regardless of this enhancement.